File: src/license/licenseCompare.ts
import { promises as fs } from 'fs';
import { parse } from 'yaml';
import { LicenseData } from '.';
import dice from 'fast-dice-coefficient';

const COPYRIGHT_REGEXP = /^(?:copyright)\s*?(?:&copy;|\(c\)|&#(?:169|xa9;)|©)\s*[0-9]{4}.*$/;

export class LicenseComparer {
    licenses: {
        name: string;
        text: string;
    }[] = [];

    async init() {
        const dir = await fs.readdir('choosealicense.com/_licenses');

        const regexp = new RegExp(COPYRIGHT_REGEXP, 'gmi');

        this.licenses = await Promise.all(dir.map(async (file) => {
            const data = await fs.readFile(`choosealicense.com/_licenses/${file}`, 'utf-8');
            const frontmatter = parse(data.split('---')[1]) as LicenseData;
            const text = data
                .split('---')[2]
                .toLowerCase()
                .replaceAll(regexp, '') // remove copyright notices
                .replaceAll(/\s+/g, '');
            return { name: frontmatter["spdx-id"], text: text };
        }));
    }

    /**
     * Returns the spdx-id of the best matching license or undefined if no match is found.
     */
    compare(license: string): string | undefined {
        const lowerCaseLicense = license.toLowerCase();

        // fast paths for common licenses
        if (lowerCaseLicense.startsWith('mit license')) {
            return 'MIT';
        }
        if (lowerCaseLicense.startsWith('GNUGENERALPUBLICLICENSEVersion3,')) {
            return 'GPL-3.0';
        }
        if (lowerCaseLicense.startsWith('GNULESSERGENERALPUBLICLICENSEVersion3,')) {
            return 'LGPL-3.0';
        }
        if (lowerCaseLicense.startsWith('GNUAFFEROGENERALPUBLICLICENSEVersion3,')) {
            return 'AGPL-3.0';
        }

        // we test if the license contains only a copyright notice like "Copyright (c) 2024 Moritz Jung"
        if (new RegExp(COPYRIGHT_REGEXP, 'gi').test(lowerCaseLicense)) {
            console.log('explicitly unlicensed', license);
            // if so we assume that the author reserves all rights
            return 'explicitly unlicensed';
        }

        const processedLicense = license
            .toLowerCase()
            .replaceAll(new RegExp(COPYRIGHT_REGEXP, 'gmi'), '') // remove copyright notices
            .replaceAll(/\s+/g, '');

        const exactMatch = this.licenses.find(x => x.text === processedLicense);
        if (exactMatch) {
            return exactMatch.name;
        }

        const scores = this.licenses.map(x => {
            return {
                name: x.name,
                score: dice(x.text, processedLicense)
            }
        });

        scores.sort((a, b) => b.score - a.score);
        const bestScore = scores[0];

        if (bestScore.score > 0.95) {
            return bestScore.name;
        }

        return undefined;
    }
}

File: data/src/plugins/license/license_compare.rs
use data_lib::license::LicenseData;
use serde_yaml;
use strsim::jaro;

use regex::Regex;

pub struct LicenseCleaned {
    name: String,
    text: String,
}

pub struct LicenseComparer {
    licenses: Vec<LicenseCleaned>,
    copyright_re: Regex,
    mit_re: Regex,
    gpl_3_re: Regex,
    lgpl_3_re: Regex,
    agpl_3_re: Regex,
}

impl Default for LicenseComparer {
    fn default() -> Self {
        LicenseComparer::new()
    }
}

impl LicenseComparer {
    pub fn new() -> Self {
        LicenseComparer {
            licenses: Vec::new(),
            copyright_re: Regex::new(
                r"^(?:copyright)\s*?(?:&copy;|\(c\)|&#(?:169|xa9;)|©)\s*[0-9]{4}.*$",
            )
            .unwrap(),
            mit_re: Regex::new(r"^mit\s+license").unwrap(),
            gpl_3_re: Regex::new(r"^gnu\s+general\s+public\s+license\s+version\s+3").unwrap(),
            lgpl_3_re: Regex::new(r"^gnu\s+lesser\s+general\s+public\s+license\s+version\s+3")
                .unwrap(),
            agpl_3_re: Regex::new(r"^gnu\s+affero\s+general\s+public\s+license\s+version\s+3")
                .unwrap(),
        }
    }

    pub fn init(&mut self) {
        let dir = std::fs::read_dir("../choosealicense.com/_licenses")
            .expect("Failed to read licenses directory");

        self.licenses = dir
            .filter_map(|entry| {
                let entry = entry.ok()?;

                let data = std::fs::read_to_string(entry.path()).ok()?;
                let parts: Vec<&str> = data.split("---").collect();
                if parts.len() <= 2 {
                    return None;
                }

                let frontmatter = serde_yaml::from_str::<LicenseData>(parts[1]).ok()?;

                Some(LicenseCleaned {
                    name: frontmatter.spdx_id,
                    text: parts[2].trim().to_lowercase().to_string(),
                })
            })
            .collect();

        println!("Loaded {} licenses", self.licenses.len());
    }

    /**
     * Returns the spdx-id of the best matching license or undefined if no match is found.
     */
    pub fn compare(&self, _plugin_id: &str, license: &str) -> Option<String> {
        let lower_case_license = license.to_lowercase();
        let lower_case_license = lower_case_license.trim();

        // println!("Comparing license: {}", lower_case_license);

        if self.mit_re.is_match(lower_case_license) {
            return Some("MIT".to_string());
        }

        if self.gpl_3_re.is_match(lower_case_license) {
            return Some("GPL-3.0".to_string());
        }

        if self.lgpl_3_re.is_match(lower_case_license) {
            return Some("LGPL-3.0".to_string());
        }

        if self.agpl_3_re.is_match(lower_case_license) {
            return Some("AGPL-3.0".to_string());
        }

        // we test if the license contains only a copyright notice like "Copyright (c) 2024 Moritz Jung"
        if self.copyright_re.is_match(lower_case_license) {
            println!("explicitly unlicensed: {license}");
            // if so we assume that the author reserves all rights
            return Some("explicitly unlicensed".to_string());
        }

        let exact_match = self.licenses.iter().find(|l| l.text == lower_case_license);
        if let Some(license) = exact_match {
            return Some(license.name.clone());
        }

        let mut scores = self
            .licenses
            .iter()
            .map(|l| {
                // jaro seems faster than normalized_levenshtein
                let score = jaro(lower_case_license, &l.text);
                (score, l.name.clone())
            })
            .collect::<Vec<_>>();

        scores.sort_by(|a, b| b.0.total_cmp(&a.0));

        // dbg!(plugin_id, &scores);

        if let Some((score, name)) = scores.first()
            && score > &0.85
        {
            return Some(name.clone());
        }

        None
    }
}
